pkg/apis/migration/migration.go

```diff
@@ -0,0 +1,351 @@
+		return "", err
+	}
+	log.Infof("successfully saved cnsvSphereVolumeMigration CR: %v", cnsvSphereVolumeMigration)
+	return fileBackingInfo.FilePath, nil
+}
+
+// SaveVolumeInfo helps create CR for given cnsVSphereVolumeMigration
+// this func also update local cache with supplied cnsVSphereVolumeMigration, after successful creation of CR
+func (volumeMigration *volumeMigration) saveVolumeInfo(ctx context.Context, cnsVSphereVolumeMigration *migrationv1alpha1.CnsVSphereVolumeMigration) error {
+	log := logger.GetLogger(ctx)
+	log.Infof("creating CR for cnsVSphereVolumeMigration: %+v", cnsVSphereVolumeMigration)
+	err := volumeMigration.k8sClient.Create(ctx, cnsVSphereVolumeMigration)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			log.Errorf("failed to create CR for cnsVSphereVolumeMigration. Error: %v", err)
+			return err
+		}
+		log.Info("CR already exists")
+		return nil
+	}
+	log.Infof("successfully created CR for cnsVSphereVolumeMigration: %+v", cnsVSphereVolumeMigration)
+	return nil
+}
+
+// DeleteVolumeInfo helps delete mapping of volumePath to VolumeID for specified volumeID
+func (volumeMigration *volumeMigration) DeleteVolumeInfo(ctx context.Context, volumeID string) error {
+	log := logger.GetLogger(ctx)
+	object := migrationv1alpha1.CnsVSphereVolumeMigration{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: volumeID,
+		},
+	}
+	err := volumeMigration.k8sClient.Delete(ctx, &object)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			log.Infof("volumeMigrationCR is already deleted for volumeID: %q", volumeID)
+			return nil
+		}
+		log.Errorf("failed delete volumeMigration CR for volumeID: %q", volumeID)
+		return err
+	}
+	return nil
+}
+
+// registerVolume takes VolumePath and helps register Volume with CNS
+// Returns VolumeID for successful registration, otherwise return error
+func (volumeMigration *volumeMigration) registerVolume(ctx context.Context, volumePath string) (string, error) {
+	log := logger.GetLogger(ctx)
+	uuid, err := uuid.NewUUID()
+	if err != nil {
+		log.Errorf("failed to generate uuid")
+		return "", err
+	}
+	re := regexp.MustCompile(`\[([^\[\]]*)\]`)
+	if !re.MatchString(volumePath) {
+		msg := fmt.Sprintf("failed to extract datastore name from in-tree volume path: %q", volumePath)
+		log.Errorf(msg)
+		return "", errors.New(msg)
+	}
+	datastoreName := re.FindAllString(volumePath, -1)[0]
+	vmdkPath := strings.TrimSpace(strings.Trim(volumePath, datastoreName))
+	datastoreName = strings.Trim(strings.Trim(datastoreName, "["), "]")
+
+	var datacenters string
+	var user string
+	var host string
+	if volumeMigration.cnsConfig == nil || len(volumeMigration.cnsConfig.VirtualCenter) == 0 {
+		msg := "could not find vcenter config"
+		log.Errorf(msg)
+		return "", errors.New(msg)
+	}
+	for key, val := range volumeMigration.cnsConfig.VirtualCenter {
+		datacenters = val.Datacenters
+		user = val.User
+		host = key
+		break
+	}
+	dataCenters := strings.Split(datacenters, ",")
```

Comment:
This could be empty as well. Don't you have to handle this empty case?

Reply:
Handling empty case by retrieving all datacenters from vCenter.

```go
        datacenterPaths := make([]string, 0)
	if datacenters != "" {
		datacenterPaths = strings.Split(datacenters, ",")
	} else {
		// Get all datacenters from vCenter
		vCenter, err := types.GetVirtualCenterInstance(ctx, &types.ConfigInfo{Cfg: volumeMigration.cnsConfig})
		if err != nil {
			log.Errorf("failed to get VirtualCenterInstance. err: %v", err)
			return "", err
		}
		err = vCenter.Connect(ctx)
		if err != nil {
			log.Errorf("failed to connect to vCenter. err: %v", err)
			return "", err
		}
		dcs, err := vCenter.GetDatacenters(ctx)
		if err != nil {
			log.Errorf("failed to get datacenters from vCenter. err: %v", err)
			return "", err
		}
		for _, dc := range dcs {
			datacenterPaths = append(datacenterPaths, dc.InventoryPath)
		}
		log.Debugf("retrieved all datacenters %v from vCenter", datacenterPaths)
	}
```